src/main/kotlin/sliv/tool/parsers/planes/ImagePlanesParser.kt

```diff
@@ -0,0 +1,108 @@
+        pixelIndex: Int,
+        imageByteDataArray: ByteArray,
+        segmentsType: ColorSegmentsType
+    ): Int {
+        val segmentsByteOffset = segmentsType.segmentsByteOffset
+
+        return convertSeparateToWholeRGB(
+            imageByteDataArray[pixelIndex + segmentsByteOffset + 2].toUByte().toInt(),
+            imageByteDataArray[pixelIndex + segmentsByteOffset + 1].toUByte().toInt(),
+            imageByteDataArray[pixelIndex + segmentsByteOffset].toUByte().toInt(),
+        )
+    }
+
+    private fun getImageByteDataArray(image: BufferedImage) = (image.data.dataBuffer as DataBufferByte).data
+
+    private fun determineImageColorSegmentsType(imagePath: String) = when (Path(imagePath).extension.lowercase()) {
+        "png" -> ColorSegmentsType.QUAD
+        "jpg", "jpeg" -> ColorSegmentsType.TRIPLE
+        else -> ColorSegmentsType.TRIPLE.also { println("Unexpected image format while parsing a plane!") }
+    }
+
+    private fun getUIDByColor(color: Int) = colorToUIDMap.getOrElse(color) {
+            println("The color to uid map does not contains value for color: $color!")
+            -1
+    }
+
+    override fun parse(filePath: String): List<Plane> {
+        val bufferedImage = ParserUtils.loadImage(filePath)
+
+        bufferedImage ?: return emptyList()
+
+        val planePoints = mutableMapOf<Long, MutableList<Point>>()
+
+        val imageByteDataArray = getImageByteDataArray(bufferedImage)
+        val colorSegmentsType: ColorSegmentsType = determineImageColorSegmentsType(filePath)
+        val colorComponentsNumber = colorSegmentsType.colorComponentsNumber
+        val imageWidth = bufferedImage.width.toShort()
+
+        for (i in imageByteDataArray.indices step colorComponentsNumber) {
+            val x = i / colorComponentsNumber % imageWidth
+            val y = i / colorComponentsNumber / imageWidth
+
+            val pixelColor = getPlanePixelColor(i, imageByteDataArray, colorSegmentsType)
+            if (pixelColor != 0) {
+                planePoints.getOrPut(getUIDByColor(pixelColor)) { mutableListOf() }.add(Point(x.toShort(), y.toShort()))
+            }
+        }
+
+        return planePoints.keys.map { Plane(it, planePoints[it] ?: emptyList()) }
+    }
+
+    override fun extractUIDs(filePath: String): List<Long> {
+        val bufferedImage = ParserUtils.loadImage(filePath)
+
+        bufferedImage ?: return emptyList()
+
+        val uids = mutableSetOf<Long>()
+
+        val imageByteDataArray = getImageByteDataArray(bufferedImage)
+        val colorSegmentsType: ColorSegmentsType = determineImageColorSegmentsType(filePath)
+        val colorComponentsNumber = colorSegmentsType.colorComponentsNumber
+        for (i in imageByteDataArray.indices step colorComponentsNumber) {
+            val pixelColor = getPlanePixelColor(i, imageByteDataArray, colorSegmentsType)
+            val uid: Long = getUIDByColor(pixelColor)
+            if (!uids.contains(uid)) {
```

Comment:
why is this check required? you have already used set

Reply:
Tests with time measurement showed a funny thing: without an initial check, the code works 20-30% slower

Reply:
it is strange, how can you explain this?

Reply:
It was a bad idea to change set to list, but there is still a problem with cheks before addition. 
I parsed uids from planes picture
````kotlin
fun main() {
    val times = mutableListOf<Long>()
    repeat(1000) {
        times.add(measureTimeMillis { ImagePlanesParser.extractUIDs("test/data/planes.png") })
    }
    val averageTime = times.average()
    val averageDeviation = times.map { abs(it) - averageTime }.average()
    println("Average time: $averageTime")
    println("Average deviation: $averageDeviation")
}
````
and got these results for variant with previous check
````
Average time: 6.043
Average deviation: -5.651479284551897E-15
````
and these are for just addition
````
Average time: 7.524
Average deviation: -8.970602038971265E-16
````
Maybe the mutable set implementation contains something other than checking for the occurrence of an element?

Reply:
It was a mistake :)